### PR TITLE
various updates to capture & materialization protocols

### PR DESCRIPTION
**Pull request by jgraettinger**
Related to factoring out materialization connector images.

### CODE REVIEW

File: capture/capture.proto [hunk 1]
```diff
@@ -175,12 +175,16 @@ message CaptureResponse {
...
+    // applied as an RFC7396 Merge Patch to the driver checkpoint
+    // of the immediately preceeding transaction (or to an empty
+    // JSON object `{}` if there is no preceeding transaction).
+    //
+    // The patched checkpoint will be persisted by the Flow
+    // runtime and returned in a future CaptureRequest.
+    bytes driver_checkpoint_merge_patch_json = 1 [
       (gogoproto.casttype) = "encoding/json.RawMessage",
-      json_name = "driverCheckpoint"
+      json_name = "driverCheckpointMergePatch"
```
Comment (saterus) [inline capture/capture.proto:187]: The idea here is that a full checkpoint will "patch" into any previous state with by overwriting everything, which while redundant, should still be correct, right?

Is there ever a circumstance that we will need to differentiate actual patches from full state checkpoints?

File: materialize/materialize.proto [hunk 2]
```diff
@@ -54,8 +54,8 @@ service Driver {
   //    - The Flow runtime client sends TransactionRequest.Open,
   //      opening the stream and requesting it be fenced from other streams.
   // :Opened:
-  //    - The driver server sends TransactionResponse.Opened after,
-  //      where supported, ensuring other stream clients are fenced.
+  //    - The driver server sends TransactionResponse.Opened after ensuring
+  //      that other raced clients are fenced (if fencing is supported).
```
Comment (saterus) [inline materialize/materialize.proto:58]: I think this is probably a bit tangential, but what exactly do you mean by "fencing" in this context? I'm not familiar with this terminology.

File: materialize/extensions.go [hunk 3]
```diff
@@ -153,6 +153,141 @@ func (m *ApplyResponse) UnmarshalJSON(b []byte) error {
...
+	if m.Commit != nil {
+		if err := m.Commit.Validate(); err != nil {
+			return pb.ExtendContext(err, "Commit")
+		}
+		count += 1
+	}
+
+	if count != 1 {
+		return pb.NewValidationError("expected one of Open, Load, Prepare, Store, Commit")
+	}
```
Comment (saterus) [inline materialize/extensions.go:191]: Is this a common go idiom for treating a struct like an enum?

**Review by saterus — APPROVED** [review]
Everything seems pretty reasonable. I was honestly expecting there to be more changes, but I think that was my misunderstanding about how far along we were with the Materialization protocol. Reviewing this helped put a lot of specifics to the concepts I've heard us discuss. :+1:

File: materialize/materialize.proto (hunk 2 above)
Comment (jgraettinger) [reply materialize/materialize.proto:58]: Terminology comes from [memory barriers](https://en.wikipedia.org/wiki/Memory_barrier). A fence is an operation which inserts a barrier such that a (distributed) actor F is assured no other actor Z (often called a "zombie") is able to race a modification of a mutually shared resource. F has "fenced" the resource from Z.

File: materialize/extensions.go (hunk 3 above)
Comment (jgraettinger) [reply materialize/extensions.go:191]: I'm not sure I've seen enough of it to call it idiomatic. Mostly I don't see it checked, and it makes me ☹️ . 

Generics are finally coming to Go, but I'd dearly love me some Rust-like algebraic enums and pattern matching....

**Pull request merged**